File: dev/node/diffSave.js
#!/usr/bin/env node
const path = require('path');
const fse = require('fs-extra');
const ArgumentParser = require('argparse').ArgumentParser;
const klawSync = require('klaw-sync');
const Sav = require('./Sav');
const diffEvent = require('./diffEvent');

/* argparse setup */
const parser = new ArgumentParser({
    addHelp: true,
    description: 'Generate a list of differences between 2 or more Gen 4-7 Pokémon save files',
});
parser.addArgument('-o', {
    metavar: 'outFile',
    help: 'File (.txt) to write diff to -- if not given, diff is written to console',
});
parser.addArgument(['-r', '--range'], {
    // action: 'append', // allow diff of multiple ranges
    nargs: 2,
    metavar: ['start', 'end'],
    help: 'Specific range to limit diff to, from start (inclusive) to end (exclusive)',
});
/* length-based range diffing
parser.addArgument('-l', {
    action: 'append',
    nargs: 2,
    metavar: ['start', 'length'],
    help: 'Specific range to limit diff to, length number of bytes beginning at start',
});
/* limit diff to current save (of split-able save file types)
parser.addArgument('-f', {
    action: 'storeTrue',
    help: 'Use full save file (of saves that can be split) in diff rather than just current save data',
});
/**/
parser.addArgument('-e', {
    action: 'storeTrue',
    help: 'Flag to include an event diff',
});
const group = parser.addMutuallyExclusiveGroup({ required: true });
group.addArgument(['-d', '--dir'], {
    metavar: 'dir',
    nargs: '+',
    help: 'Space separated list of directories (folders) containing files to diff',
});
group.addArgument(['-s', '--saves'], {
    nargs: '+',
    help: 'Space separated list of save files to diff',
});

/* Global variables */
let params;
let diffs = 0;

/**
 * @param {Number} value
 * @param {Number} padTo
 * @returns {String}
 */
function toHexStr(value, padTo) {
    let paddedHex = value.toString(16);
    while (paddedHex.length < padTo) {
        paddedHex = `0${paddedHex}`;
    }
    return `0x${paddedHex.slice(-padTo)}`;
}

/**
 * @param {Sav[]} saves
 * @param {Object} config
 * @param {Boolean} config.event
 * @param {Number[]} config.range
 * @param {Boolean} config.isSplit
 * @returns {String}
 */
function generateDiff(saves, config) {
    const diff = [];

    // construct diff header
    const tableHeader = ['\nOffset   '];
    diff.push('Diff of saves from the following file(s):\n\n');
    let ofs = {};
    if (saves[0].generation === 4) {
        ofs.general = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'general');
        ofs.storage = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'storage');
        ofs.hof = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'hof');
    } else if (saves[0].generation === 5) {
        ofs.save = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'save');
        ofs.extra = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'extra');
    }
    saves.forEach((v, i) => {
        diff.push(`Save ${i + 1}: ${v.fileName}${config.isSplit ? ' (' + v.contents + ')' : ''}\n`);

        // report block offsets of split saves
        if (config.isSplit) {
            if (v.generation === 4) { // split Gen 4 save
                diff.push(`    General: ${ofs.general.map(n => toHexStr(n + v.active.general * 0x40000, 5)).join(':')}\n`);
                diff.push(`    Storage: ${ofs.storage.map(n => toHexStr(n + v.active.storage * 0x40000, 5)).join(':')}\n`);
                diff.push(`    HoF:     ${ofs.hof.map(n => toHexStr(n + v.active.hof * 0x40000, 5)).join(':')}\n`);
            } else if (v.verGroupAbbr === 'BW') { // split BW save
                diff.push(`    Save:  ${ofs.save.map(n => toHexStr(n + v.active.save * 0x24000, 5)).join(':')}\n`);
                diff.push(`    Extra: ${ofs.extra.map(n => toHexStr(n, 5)).join(':')}\n`);
            } else if (v.verGroupAbbr === 'B2W2') { // split B2W2 save
                diff.push(`    Save:  ${ofs.save.map(n => toHexStr(n + v.active.save * 0x26000, 5)).join(':')}\n`);
                diff.push(`    Extra: ${ofs.extra.map(n => toHexStr(n, 5)).join(':')}\n`);
            } else {
                console.log('How did you split this input?');
            }
        }

        let digits = 10 + (Math.floor(Math.log10(i + 1)) !== Math.floor(Math.log10(i)));
        tableHeader.push(`    Save ${i + 1}`.slice(-digits));
    });

    if (saves[0].generation === 4 && config.isSplit) {
        diff.push('\nNote: HoF block of Save 1 may be more recent than the HoF block of Save 2\n');
    }

    // event diffing
    if (config.event) {
        let eDiff = diffEvent(saves);
        if (eDiff == '') {
            console.log('Event diff failed.');
        } else {
            diffs += eDiff.slice(0, 25) === 'Event flags and constants' ? 0 : 1;
            diff.push("\n", eDiff);
        }
    }

    // range slicing
    let slicedData = saves.map(v => v.data);
    // let slicedData = saves.map(v => config.f ? v.data : v.getCurrent());
    let sliceStart = 0;
    if (config.range) {
        sliceStart = +config.range[0];
        slicedData = slicedData.map(v => v.slice(+config.range[0], +config.range[1]));
        diff.push(`\n\nDiff range: ${config.range[0]}-${config.range[1]}`);
    }

    // construct diff body
    diff.push(`\n\n${tableHeader.join('')}\n`);
    slicedData[0].forEach((d, n) => {
        const offsetDiff = {
            offset: toHexStr(n + sliceStart, 5),
            values: [],
        };
        if (config.isSplit && saves[0].generation === 5) {
            if (n > ofs.save[1]) {
                offsetDiff.offset = toHexStr(n + sliceStart + ofs.save[1] + 1);
            }
        }
        slicedData.forEach((v, i) => {
            offsetDiff.values[i] = toHexStr(v[n], 2);
        });
        const match = offsetDiff.values.every((v, i, a) => v === a[0]);
        if (!match) {
            diffs += 1;
            diff.push(`${offsetDiff.offset}      ${offsetDiff.values.join('      ')}\n`);
        }
    });

    return diff.join('');
}

/**
 * @param {Object} params
 * @param {String[]} params.saves
 * @param {String} params.o (Optional)
 * @param {String[]} params.range (Optional)
 * @param {Boolean} params.e (Optional)
 */
function main(params) {
    const config = {
        range: params.range,
        event: params.e,
    };

    console.log('Loading saves...');
    const saves = params.saves.map(v => new Sav(v)).filter(v => v.version !== -1);

    console.log('Checking compatibility of saves...');
    let compatFailReason = Sav.checkCompat(saves);
    if (compatFailReason !== '') {
        console.log(`Save compatibility check failed.\nReason: "${compatFailReason}"\nExiting without diffing.`);
        return;
    }

    // try to split lone save
    if (saves.length === 1) {
        if (!saves[0].canSplit()) {
            console.log(`Cannot split a single ${saves[0].verGroupAbbr} save file. Exiting without diffing.`);
            return;
        }
        // split lone save
        Sav.splitSave(saves, 0, true);
        config.isSplit = true;
    }

    console.log('Generating diff...');
    const diff = generateDiff(saves, config);

    // output diff
    if (diffs === 0) {
        console.log(`Saves were identical.${params.o ? " " + params.o + " not written to." : ""}`);
    } else if (params.o) {
        let outFile = params.o;
        outFile = `${outFile}${outFile.slice(-4) === '.txt' ? '' : '.txt'}`;
        fse.outputFileSync(outFile, diff);
        console.log(`Diff written to ${outFile} successfully.`);
    } else {
        console.log(`\n\n${diff}`);
    }
}

// execute if called directly from command line
if (module === require.main) {
    params = parser.parseArgs();
    if (params.dir) {
        if (!params.saves) {
            params.saves = [];
        }
        params.dir.forEach(d => {
            const paths = klawSync(d, { nodir: true });
            const fileNames = paths.map(v => path.relative('', v.path));
            Array.prototype.push.apply(params.saves, fileNames);
        })
    }
    main(params);
}

File: dev/python/diffSave.py
#!/usr/bin/env python3
import math
import argparse
import os

from Sav import Sav
from diffEvent import generateDiff as diffEvent

# argparse setup
parser = argparse.ArgumentParser(
    description='Generate a list of differences between 2 or more Gen 4-7 Pokémon saves')
parser.add_argument('-o', metavar='outFile',
                    help='File (.txt) to write diff to -- if not given, diff is written to console')
parser.add_argument('-r', '--range', nargs=2, metavar=('start', 'end'),
                    help='Specific range to limit diff to, from start (included) to end (excluded)')
# Length-based range diffing
# parser.add_argument('-l', action='append', nargs=2, metavar=('start', 'length'),
#                     help='Specific range to limit diff to, length number of bytes beginning at start')
# Diff full file rather than just current save data (of split-able save file types)
# parser.add_argument('-f', action='store_true',
#                     help='Use full save file (of saves that can be split) in diff rather than just current save data')
parser.add_argument('-e', action='store_true',
                    help='Flag to include an event diff')
group = parser.add_mutually_exclusive_group(required=True)
group.add_argument('-d', '--dir', metavar='dir', nargs='+',
                   help='Directory (folder) containing files to diff')
group.add_argument('-s', '--saves', nargs='+',
                   help='Space separated list of save files to diff')

# Global variables
diffs = 0


def _toPaddedHexString(value, padTo):
    paddedHex = hex(value)[2:]
    while len(paddedHex) < padTo:
        paddedHex = "0" + paddedHex
    return "0x" + paddedHex[-padTo:]


def generateDiff(saves, config):
    diff = []
    global diffs

    ofs = {
        'general': [0],
        'storage': [0],
        'hof': [0],
        'save': [0],
        'extra': [0]
    }
    if saves[0].generation == 4:
        ofs['general'] = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'general')
        ofs['storage'] = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'storage')
        ofs['hof'] = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'hof')
    elif saves[0].generation == 5:
        ofs['save'] = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'save')
        ofs['extra'] = Sav.getBlockOffsets(saves[0].verGroupAbbr, 'extra')

    # construct diff header
    tableHeader = ["\nOffset   "]
    diff.append("Diff of saves from the following file(s):\n\n")
    for i in range(len(saves)):
        contents = '' if saves[i].contents == 'full' else ' (%s)' % saves[i].contents
        diff.append("Save %d: %s%s\n" % (i + 1, saves[i].fileName, contents))

        if config['isSplit']:
            if saves[i].generation == 4:
                compOfs = [_toPaddedHexString(n + saves[i].active['general'] * 0x40000, 5) for n in ofs['general']]
                diff.append("    General: %s\n" % ':'.join(compOfs))
                compOfs = [_toPaddedHexString(n + saves[i].active['storage'] * 0x40000, 5) for n in ofs['storage']]
                diff.append("    Storage: %s\n" % ':'.join(compOfs))
                compOfs = [_toPaddedHexString(n + saves[i].active['hof'] * 0x40000, 5) for n in ofs['hof']]
                diff.append("    HoF:     %s\n" % ':'.join(compOfs))
            elif saves[i].verGroupAbbr == 'BW':
                compOfs = [_toPaddedHexString(n + saves[i].active['save'] * 0x24000, 5) for n in ofs['save']]
                diff.append("    Save:  %s\n" % ':'.join(compOfs))
                compOfs = [_toPaddedHexString(n + saves[i].active['extra'] * 0x24000, 5) for n in ofs['extra']]
                diff.append("    Extra: %s\n" % ':'.join(compOfs))
            elif saves[i].verGroupAbbr == 'B2W2':
                compOfs = [_toPaddedHexString(n + saves[i].active['save'] * 0x26000, 5) for n in ofs['save']]
                diff.append("    Save:  %s\n" % ':'.join(compOfs))
                compOfs = [_toPaddedHexString(n + saves[i].active['extra'] * 0x26000, 5) for n in ofs['extra']]
                diff.append("    Extra: %s\n" % ':'.join(compOfs))
            else:
                print('How did you split this input?')

        digStr = "    Save %d" % (i + 1)
        digits = 10
        if i % 10 == 9:
            digits += (math.floor(math.log10(i + 1)) != math.floor(math.log10(i)))
        tableHeader.append(digStr[-digits:])

    if saves[0].generation == 4 and config['isSplit']:
        diff.append(
            "\nNote: HoF block of Save 1 may be more recent than the HoF block of Save 2\n")

    # event diffing (function call?) goes here
    if config['event']:
        eDiff = diffEvent(saves)
        if eDiff == '':
            print('Event diff failed.')
        else:
            diffs += 0 if eDiff[:25] == 'Event flags and constants' else 1
            diff.append("\n%s" % eDiff)

    # range slicing
    slicedData = [saves[i].data for i in range(len(saves))]
    sliceStart = 0
    if config['range'] != None:
        sliceStart = int(config['range'][0], 0)
        for i in range(len(slicedData)):
            slicedData[i] = slicedData[i][sliceStart:int(config['range'][1], 0)]
        diff.append("\n\nDiff range: %s-%s" % (config['range'][0], config['range'][1]))

    # construct diff body
    diff.append("\n\n%s\n" % ''.join(tableHeader))
    for n in range(len(slicedData[0])):
        offsetDiff = {
            'offset': _toPaddedHexString(n + sliceStart, 5),
            'values': []
        }
        if config['isSplit'] and saves[0].generation == 5:
            if n > ofs['save'][1]:
                offsetDiff['offset'] = _toPaddedHexString(n + sliceStart + ofs['save'][1] + 1, 2)
            pass
        offsetDiff['values'] = [_toPaddedHexString(
            slicedData[i][n], 2) for i in range(len(slicedData))]
        for i in range(len(offsetDiff['values'])):
            if offsetDiff['values'][i] != offsetDiff['values'][0]:
                diffs += 1
                diff.append("%s      %s\n" % (
                    offsetDiff['offset'], '      '.join(offsetDiff['values'])))
                break

    # join diff pieces
    return ''.join(diff)


def main(params):
    config = {
        'range': params.range,
        'event': params.e,
        'isSplit': False
    }

    print("Loading saves...")
    saves = [Sav(f) for f in params.saves]
    saves = [s for s in saves if s.version != -1]

    print("Checking compatibility of saves...")
    compatFailReason = Sav.checkCompat(saves)
    if compatFailReason != '':
        print("Save compatibility check failed.\nReason: \"%s\"\nExiting without diffing." %
              compatFailReason)
        return

    # split lone Gen 4 save
    if len(saves) == 1:
        if not saves[0].canSplit():
            print(
                "Cannot split a single %s save file. Exiting without diffing." % saves[0].verGroupAbbr)
            return
        # split lone save
        Sav.splitSave(saves, 0, True)
        config['isSplit'] = True

    print("Generating diff...")
    diff = generateDiff(saves, config)

    # output diff
    if diffs == 0:
        print("Saves were identical.%s" %
              ("" if params.o == None else " %s not written to." % params.o))
    elif params.o != None:
        outDir = ''
        try:
            outDir = os.path.dirname(params.o[0])
            os.makedirs(outDir, exist_ok=True)
        except:
            pass
        outFile = "%s%s" % (
            params.o, "" if params.o[-4:] == ".txt" else ".txt")
        with open(outFile, 'w') as f:
            f.write(diff)
        print("Diff written to %s successfully" % outFile)
    else:
        print("\n\n%s" % diff)


# execute if called directly from command line
if __name__ == '__main__':
    params = parser.parse_args()
    if params.dir != None:
        if params.saves == None:
            params.saves = []
        for d in range(len(params.dir)):
            dirFiles = []
            for path, _, files in os.walk(params.dir[d]):
                for fullname in files:
                    dirFiles.append(os.path.normpath(
                        os.path.join(path, fullname)))
            dirFiles.sort(key=str.lower)
            params.saves.extend(dirFiles)
    main(params)
